notebooks/hello_world_thunderfx.ipynb

```diff
@@ -0,0 +1,279 @@
+{
```

Comment:
<p>This actually seems like a very high absolute deviation! Did the <code>assert_close</code> call fail?</p><p> </p><p>We naively expect bfloat16 absolute numerical inconsistencies to be less than 1e-5 (see <a href="https://pytorch.org/docs/stable/testing.html#torch.testing.assert_close" target="_blank">https://pytorch.org/docs/stable/testing.html#torch.testing.assert_close</a>), although the actual deviation is dependent on the function being tested (the number of operations in the function, for example, changes our expectations for numerical consistency). In this case I wouldn't expect such a large deviation, but I could be mistaken -- I wonder if we should try to determine what's happening?</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Lightning-AI/lightning-thunder/pull/1524/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Reply:
I think the inconsistencies may come from the accumulation of operators and that the operators are decomposed. torch.compile and eager also have similar deviation as thunderfx compared to eager. I noticed that in the zero-to-thunder notebook, we compare the absolute deviation for Llama-2-7b-hf(16layers,bf16), is about 0.03125.
Here are some results on Llama-3-8B(2 layers, block_size=1024, batch_dim=1)
```
Torchcompile--eager -- f32
Mismatched elements: 113611144 / 131334144 (86.5%)
Greatest absolute difference: 0.0004018545150756836 at index (0, 685, 45706) (up to 1e-05 allowed)

Thunderfx--eager fp32
Mismatched elements: 113771104 / 131334144 (86.6%)
Greatest absolute difference: 0.00045102834701538086 at index (0, 5, 117396) (up to 1e-05 allowed)

Thunderfx--torch.compile fp32
Mismatched elements: 113077014 / 131334144 (86.1%)
Greatest absolute difference: 0.0003840327262878418 at index (0, 372, 64716) (up to 1e-05 allowed)
```
with bf16 the Greatest absolute difference is a bit larger：
`Torchcompile--eager : Thunderfx--eager  => 0.013671875 : 0.015625`

Also I saved the reproducer function and compare the result of it directly:
Greatest absolute difference(fp32)
`Thunderfx -- eager : Torch.compile -- eager  =>  0.00022476911544799805 :  0.00021338462829589844`
P.S: the above results have set the seed and I checked for each executor, if we run the forward with the same input twice the results are the same, 
<details>
<summary>The repro function</summary>

```
from math import inf
from math import nan
NoneType = type(None)
import torch
from torch import device
import torch.fx._pytree as fx_pytree
import torch.utils._pytree as pytree
from functools import partial
import thunder
from thunder.transforms.cudagraph import CUDAGraphTransform
from thunder.dev_utils.nvtx_profile_transform import NvtxProfileTransform


def set_seed(seed):
    import random
    import numpy as np
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    random.seed(seed)
    np.random.seed(seed)
    # torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False
    # torch.use_deterministic_algorithms(True)

def test_graph0_thunder_0():
    class DynamoModule(torch.nn.Module):
      def forward(self, l_self_buffers_cos_ : torch.Tensor, l_self_buffers_sin_ : torch.Tensor, l_idx_ : torch.Tensor, l_self_modules_transformer_modules_wte_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_h_modules_0_modules_norm_1_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_h_modules_0_modules_attn_modules_attn_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_h_modules_0_modules_attn_modules_proj_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_h_modules_0_modules_norm_2_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_h_modules_0_modules_mlp_modules_fc_1_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_h_modules_0_modules_mlp_modules_fc_2_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_h_modules_0_modules_mlp_modules_proj_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_h_modules_1_modules_norm_1_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_h_modules_1_modules_attn_modules_attn_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_h_modules_1_modules_attn_modules_proj_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_h_modules_1_modules_norm_2_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_h_modules_1_modules_mlp_modules_fc_1_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_h_modules_1_modules_mlp_modules_fc_2_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_h_modules_1_modules_mlp_modules_proj_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_transformer_modules_ln_f_parameters_weight_ : torch.nn.parameter.Parameter, l_self_modules_lm_head_parameters_weight_ : torch.nn.parameter.Parameter):
          cos = l_self_buffers_cos_[slice(None, 1024, None)];  l_self_buffers_cos_ = None
          sin = l_self_buffers_sin_[slice(None, 1024, None)];  l_self_buffers_sin_ = None
          x = torch.nn.functional.embedding(l_idx_, l_self_modules_transformer_modules_wte_parameters_weight_, None, None, 2.0, False, False);  l_idx_ = l_self_modules_transformer_modules_wte_parameters_weight_ = None
          x_1 = x.float()
          mul = x_1 * x_1
          norm_x = torch.mean(mul, dim = -1, keepdim = True);  mul = None
          add = norm_x + 1e-05;  norm_x = None
          rsqrt = torch.rsqrt(add);  add = None
          x_normed = x_1 * rsqrt;  x_1 = rsqrt = None
          float_2 = l_self_modules_transformer_modules_h_modules_0_modules_norm_1_parameters_weight_.float();  l_self_modules_transformer_modules_h_modules_0_modules_norm_1_parameters_weight_ = None
          mul_2 = x_normed * float_2;  x_normed = float_2 = None
          x_normed_1 = mul_2.to(dtype = torch.float32);  mul_2 = None
          qkv = torch._C._nn.linear(x_normed_1, l_self_modules_transformer_modules_h_modules_0_modules_attn_modules_attn_parameters_weight_, None);  x_normed_1 = l_self_modules_transformer_modules_h_modules_0_modules_attn_modules_attn_parameters_weight_ = None
          qkv_1 = qkv.view(1, 1024, 8, 6, 128);  qkv = None
          qkv_2 = qkv_1.permute(0, 2, 3, 1, 4);  qkv_1 = None
          split = qkv_2.split((4, 1, 1), dim = 2);  qkv_2 = None
          q = split[0]
          k = split[1]
          v = split[2];  split = None
          k_1 = k.expand(1, 8, 4, 1024, 128);  k = None
          v_1 = v.expand(1, 8, 4, 1024, 128);  v = None
          q_1 = q.reshape(1, -1, 1024, 128);  q = None
          k_2 = k_1.reshape(1, -1, 1024, 128);  k_1 = None
          v_2 = v_1.reshape(1, -1, 1024, 128);  v_1 = None
          getitem_5 = q_1[(Ellipsis, slice(None, 128, None))]
          x1 = getitem_5[(Ellipsis, slice(None, 64, None))]
          x2 = getitem_5[(Ellipsis, slice(64, None, None))]
          neg = -x2;  x2 = None
          rotated = torch.cat((neg, x1), dim = -1);  neg = x1 = None
          mul_3 = getitem_5 * cos;  getitem_5 = None
          mul_4 = rotated * sin;  rotated = None
          roped = mul_3 + mul_4;  mul_3 = mul_4 = None
          q_roped = roped.to(dtype = torch.float32);  roped = None
          getitem_8 = k_2[(Ellipsis, slice(None, 128, None))]
          x1_1 = getitem_8[(Ellipsis, slice(None, 64, None))]
          x2_1 = getitem_8[(Ellipsis, slice(64, None, None))]
          neg_1 = -x2_1;  x2_1 = None
          rotated_1 = torch.cat((neg_1, x1_1), dim = -1);  neg_1 = x1_1 = None
          mul_5 = getitem_8 * cos;  getitem_8 = None
          mul_6 = rotated_1 * sin;  rotated_1 = None
          roped_1 = mul_5 + mul_6;  mul_5 = mul_6 = None
          k_roped = roped_1.to(dtype = torch.float32);  roped_1 = None
          getitem_11 = q_1[(Ellipsis, slice(128, None, None))];  q_1 = None
          q_2 = torch.cat((q_roped, getitem_11), dim = -1);  q_roped = getitem_11 = None
          getitem_12 = k_2[(Ellipsis, slice(128, None, None))];  k_2 = None
          k_3 = torch.cat((k_roped, getitem_12), dim = -1);  k_roped = getitem_12 = None
          y = torch._C._nn.scaled_dot_product_attention(q_2, k_3, v_2, attn_mask = None, dropout_p = 0.0, scale = 0.08838834764831843, is_causal = True);  q_2 = k_3 = v_2 = None
          y_1 = y.transpose(1, 2);  y = None
          y_2 = y_1.reshape(1, 1024, 4096);  y_1 = None
          attention_output = torch._C._nn.linear(y_2, l_self_modules_transformer_modules_h_modules_0_modules_attn_modules_proj_parameters_weight_, None);  y_2 = l_self_modules_transformer_modules_h_modules_0_modules_attn_modules_proj_parameters_weight_ = None
          x_2 = attention_output + x;  attention_output = x = None
          x_3 = x_2.float()
          mul_7 = x_3 * x_3
          norm_x_1 = torch.mean(mul_7, dim = -1, keepdim = True);  mul_7 = None
          add_4 = norm_x_1 + 1e-05;  norm_x_1 = None
          rsqrt_1 = torch.rsqrt(add_4);  add_4 = None
          x_normed_2 = x_3 * rsqrt_1;  x_3 = rsqrt_1 = None
          float_4 = l_self_modules_transformer_modules_h_modules_0_modules_norm_2_parameters_weight_.float();  l_self_modules_transformer_modules_h_modules_0_modules_norm_2_parameters_weight_ = None
          mul_9 = x_normed_2 * float_4;  x_normed_2 = float_4 = None
          to_3 = mul_9.to(dtype = torch.float32);  mul_9 = None
          x_fc_1 = torch._C._nn.linear(to_3, l_self_modules_transformer_modules_h_modules_0_modules_mlp_modules_fc_1_parameters_weight_, None);  l_self_modules_transformer_modules_h_modules_0_modules_mlp_modules_fc_1_parameters_weight_ = None
          x_fc_2 = torch._C._nn.linear(to_3, l_self_modules_transformer_modules_h_modules_0_modules_mlp_modules_fc_2_parameters_weight_, None);  to_3 = l_self_modules_transformer_modules_h_modules_0_modules_mlp_modules_fc_2_parameters_weight_ = None
          silu = torch.nn.functional.silu(x_fc_1);  x_fc_1 = None
          x_4 = silu * x_fc_2;  silu = x_fc_2 = None
          linear_4 = torch._C._nn.linear(x_4, l_self_modules_transformer_modules_h_modules_0_modules_mlp_modules_proj_parameters_weight_, None);  x_4 = l_self_modules_transformer_modules_h_modules_0_modules_mlp_modules_proj_parameters_weight_ = None
          x_5 = linear_4 + x_2;  linear_4 = x_2 = None
          x_6 = x_5.float()
          mul_11 = x_6 * x_6
          norm_x_2 = torch.mean(mul_11, dim = -1, keepdim = True);  mul_11 = None
          add_6 = norm_x_2 + 1e-05;  norm_x_2 = None
          rsqrt_2 = torch.rsqrt(add_6);  add_6 = None
          x_normed_3 = x_6 * rsqrt_2;  x_6 = rsqrt_2 = None
          float_6 = l_self_modules_transformer_modules_h_modules_1_modules_norm_1_parameters_weight_.float();  l_self_modules_transformer_modules_h_modules_1_modules_norm_1_parameters_weight_ = None
          mul_13 = x_normed_3 * float_6;  x_normed_3 = float_6 = None
          x_normed_4 = mul_13.to(dtype = torch.float32);  mul_13 = None
          qkv_3 = torch._C._nn.linear(x_normed_4, l_self_modules_transformer_modules_h_modules_1_modules_attn_modules_attn_parameters_weight_, None);  x_normed_4 = l_self_modules_transformer_modules_h_modules_1_modules_attn_modules_attn_parameters_weight_ = None
          qkv_4 = qkv_3.view(1, 1024, 8, 6, 128);  qkv_3 = None
          qkv_5 = qkv_4.permute(0, 2, 3, 1, 4);  qkv_4 = None
          split_1 = qkv_5.split((4, 1, 1), dim = 2);  qkv_5 = None
          q_3 = split_1[0]
          k_4 = split_1[1]
          v_3 = split_1[2];  split_1 = None
          k_5 = k_4.expand(1, 8, 4, 1024, 128);  k_4 = None
          v_4 = v_3.expand(1, 8, 4, 1024, 128);  v_3 = None
          q_4 = q_3.reshape(1, -1, 1024, 128);  q_3 = None
          k_6 = k_5.reshape(1, -1, 1024, 128);  k_5 = None
          v_5 = v_4.reshape(1, -1, 1024, 128);  v_4 = None
          getitem_16 = q_4[(Ellipsis, slice(None, 128, None))]
          x1_2 = getitem_16[(Ellipsis, slice(None, 64, None))]
          x2_2 = getitem_16[(Ellipsis, slice(64, None, None))]
          neg_2 = -x2_2;  x2_2 = None
          rotated_2 = torch.cat((neg_2, x1_2), dim = -1);  neg_2 = x1_2 = None
          mul_14 = getitem_16 * cos;  getitem_16 = None
          mul_15 = rotated_2 * sin;  rotated_2 = None
          roped_2 = mul_14 + mul_15;  mul_14 = mul_15 = None
          q_roped_1 = roped_2.to(dtype = torch.float32);  roped_2 = None
          getitem_19 = k_6[(Ellipsis, slice(None, 128, None))]
          x1_3 = getitem_19[(Ellipsis, slice(None, 64, None))]
          x2_3 = getitem_19[(Ellipsis, slice(64, None, None))]
          neg_3 = -x2_3;  x2_3 = None
          rotated_3 = torch.cat((neg_3, x1_3), dim = -1);  neg_3 = x1_3 = None
          mul_16 = getitem_19 * cos;  getitem_19 = cos = None
          mul_17 = rotated_3 * sin;  rotated_3 = sin = None
          roped_3 = mul_16 + mul_17;  mul_16 = mul_17 = None
          k_roped_1 = roped_3.to(dtype = torch.float32);  roped_3 = None
          getitem_22 = q_4[(Ellipsis, slice(128, None, None))];  q_4 = None
          q_5 = torch.cat((q_roped_1, getitem_22), dim = -1);  q_roped_1 = getitem_22 = None
          getitem_23 = k_6[(Ellipsis, slice(128, None, None))];  k_6 = None
          k_7 = torch.cat((k_roped_1, getitem_23), dim = -1);  k_roped_1 = getitem_23 = None
          y_3 = torch._C._nn.scaled_dot_product_attention(q_5, k_7, v_5, attn_mask = None, dropout_p = 0.0, scale = 0.08838834764831843, is_causal = True);  q_5 = k_7 = v_5 = None
          y_4 = y_3.transpose(1, 2);  y_3 = None
          y_5 = y_4.reshape(1, 1024, 4096);  y_4 = None
          attention_output_1 = torch._C._nn.linear(y_5, l_self_modules_transformer_modules_h_modules_1_modules_attn_modules_proj_parameters_weight_, None);  y_5 = l_self_modules_transformer_modules_h_modules_1_modules_attn_modules_proj_parameters_weight_ = None
          x_7 = attention_output_1 + x_5;  attention_output_1 = x_5 = None
          x_8 = x_7.float()
          mul_18 = x_8 * x_8
          norm_x_3 = torch.mean(mul_18, dim = -1, keepdim = True);  mul_18 = None
          add_10 = norm_x_3 + 1e-05;  norm_x_3 = None
          rsqrt_3 = torch.rsqrt(add_10);  add_10 = None
          x_normed_5 = x_8 * rsqrt_3;  x_8 = rsqrt_3 = None
          float_8 = l_self_modules_transformer_modules_h_modules_1_modules_norm_2_parameters_weight_.float();  l_self_modules_transformer_modules_h_modules_1_modules_norm_2_parameters_weight_ = None
          mul_20 = x_normed_5 * float_8;  x_normed_5 = float_8 = None
          to_7 = mul_20.to(dtype = torch.float32);  mul_20 = None
          x_fc_3 = torch._C._nn.linear(to_7, l_self_modules_transformer_modules_h_modules_1_modules_mlp_modules_fc_1_parameters_weight_, None);  l_self_modules_transformer_modules_h_modules_1_modules_mlp_modules_fc_1_parameters_weight_ = None
          x_fc_4 = torch._C._nn.linear(to_7, l_self_modules_transformer_modules_h_modules_1_modules_mlp_modules_fc_2_parameters_weight_, None);  to_7 = l_self_modules_transformer_modules_h_modules_1_modules_mlp_modules_fc_2_parameters_weight_ = None
          silu_1 = torch.nn.functional.silu(x_fc_3);  x_fc_3 = None
          x_9 = silu_1 * x_fc_4;  silu_1 = x_fc_4 = None
          linear_9 = torch._C._nn.linear(x_9, l_self_modules_transformer_modules_h_modules_1_modules_mlp_modules_proj_parameters_weight_, None);  x_9 = l_self_modules_transformer_modules_h_modules_1_modules_mlp_modules_proj_parameters_weight_ = None
          x_10 = linear_9 + x_7;  linear_9 = x_7 = None
          x_11 = x_10.float();  x_10 = None
          mul_22 = x_11 * x_11
          norm_x_4 = torch.mean(mul_22, dim = -1, keepdim = True);  mul_22 = None
          add_12 = norm_x_4 + 1e-05;  norm_x_4 = None
          rsqrt_4 = torch.rsqrt(add_12);  add_12 = None
          x_normed_6 = x_11 * rsqrt_4;  x_11 = rsqrt_4 = None
          float_10 = l_self_modules_transformer_modules_ln_f_parameters_weight_.float();  l_self_modules_transformer_modules_ln_f_parameters_weight_ = None
          mul_24 = x_normed_6 * float_10;  x_normed_6 = float_10 = None
          x_12 = mul_24.to(dtype = torch.float32);  mul_24 = None
          x_13 = torch._C._nn.linear(x_12, l_self_modules_lm_head_parameters_weight_, None);  x_12 = l_self_modules_lm_head_parameters_weight_ = None
          return x_13
    set_seed(10)   
    inputs = [
        torch.testing.make_tensor((1024, 128), dtype=torch.float32,  device='cuda:0', requires_grad=False, low=-1.0, high=1.0,).as_strided((1024, 128), (128, 1)),
        torch.testing.make_tensor((1024, 128), dtype=torch.float32,  device='cuda:0', requires_grad=False, low=-0.9999999403953552, high=1.0,).as_strided((1024, 128), (128, 1)),
        torch.testing.make_tensor((1, 1024), dtype=torch.int64,  device='cuda:0', requires_grad=False, low=0, high=254,).as_strided((1, 1024), (1024, 1)),
        torch.testing.make_tensor((128256, 4096), dtype=torch.float32,  device='cuda:0', requires_grad=True, low=-5.638088226318359, high=6.174197196960449,).as_strided((128256, 4096), (4096, 1)),
        torch.full((4096,), 1.0, dtype=torch.float32, device='cuda:0', requires_grad=True, layout=torch.strided).as_strided((4096,), (1,)),
        torch.testing.make_tensor((6144, 4096), dtype=torch.float32,  device='cuda:0', requires_grad=True, low=-0.01562499813735485, high=0.01562499813735485,).as_strided((6144, 4096), (4096, 1)),
        torch.testing.make_tensor((4096, 4096), dtype=torch.float32,  device='cuda:0', requires_grad=True, low=-0.015624997206032276, high=0.015624996274709702,).as_strided((4096, 4096), (4096, 1)),
        torch.full((4096,), 1.0, dtype=torch.float32, device='cuda:0', requires_grad=True, layout=torch.strided).as_strided((4096,), (1,)),
        torch.testing.make_tensor((14336, 4096), dtype=torch.float32,  device='cuda:0', requires_grad=True, low=-0.015625, high=0.01562499813735485,).as_strided((14336, 4096), (4096, 1)),
        torch.testing.make_tensor((14336, 4096), dtype=torch.float32,  device='cuda:0', requires_grad=True, low=-0.015625, high=0.01562499813735485,).as_strided((14336, 4096), (4096, 1)),
        torch.testing.make_tensor((4096, 14336), dtype=torch.float32,  device='cuda:0', requires_grad=True, low=-0.008351913653314114, high=0.008351912721991539,).as_strided((4096, 14336), (14336, 1)),
        torch.full((4096,), 1.0, dtype=torch.float32, device='cuda:0', requires_grad=True, layout=torch.strided).as_strided((4096,), (1,)),
        torch.testing.make_tensor((6144, 4096), dtype=torch.float32,  device='cuda:0', requires_grad=True, low=-0.015625, high=0.01562499813735485,).as_strided((6144, 4096), (4096, 1)),
        torch.testing.make_tensor((4096, 4096), dtype=torch.float32,  device='cuda:0', requires_grad=True, low=-0.015625, high=0.015624996274709702,).as_strided((4096, 4096), (4096, 1)),
        torch.full((4096,), 1.0, dtype=torch.float32, device='cuda:0', requires_grad=True, layout=torch.strided).as_strided((4096,), (1,)),
        torch.testing.make_tensor((14336, 4096), dtype=torch.float32,  device='cuda:0', requires_grad=True, low=-0.015625, high=0.01562499813735485,).as_strided((14336, 4096), (4096, 1)),
        torch.testing.make_tensor((14336, 4096), dtype=torch.float32,  device='cuda:0', requires_grad=True, low=-0.015625, high=0.01562499813735485,).as_strided((14336, 4096), (4096, 1)),
        torch.testing.make_tensor((4096, 14336), dtype=torch.float32,  device='cuda:0', requires_grad=True, low=-0.008351913653314114, high=0.008351912721991539,).as_strided((4096, 14336), (14336, 1)),
        torch.full((4096,), 1.0, dtype=torch.float32, device='cuda:0', requires_grad=True, layout=torch.strided).as_strided((4096,), (1,)),
        torch.testing.make_tensor((128256, 4096), dtype=torch.float32,  device='cuda:0', requires_grad=True, low=-0.015625, high=0.01562499813735485,).as_strided((128256, 4096), (4096, 1)),
        
    ]
    compiled = thunder.jit(DynamoModule(), )
    out1 = compiled(*inputs)
    # compiled = thunder.jit(DynamoModule(), )
    # out2 = compiled(*inputs)
    # out2 = torch.compile(DynamoModule())(*inputs)
    # out1 = torch.compile(DynamoModule())(*inputs)
    # out1 = DynamoModule()(*inputs)
    out2 = DynamoModule()(*inputs)
    # print(out1.eq(out2).all())
    torch.testing.assert_close(out1, out2)

test_graph0_thunder_0()
```

</details>

Reply:
<p>This is a fantastic analysis. I think we just need to add a comment on the deviation, then. Maybe something like:</p><p> </p><p>"Thunder's optimized kernels may produce slightly different results than other kernels, but the differences shouldn't be significant in practice."</p><p> </p><p>What do you think? </p>